File: Compose-Bookshelf-App/app/src/main/java/com/example/bookshelfapp/data/BooksRepository.kt
package com.example.bookshelfapp.data

import com.example.bookshelfapp.model.GoogleBookApiModel
import com.example.bookshelfapp.model.GoogleBookRetrieveApiModel
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BooksRepository @Inject constructor(
    private val booksDataSource: BooksDataSource,
    private val thumbnailDataSource: BookThumbnailDataSource
) {
    suspend fun getBooks(searchTerms: String): GoogleBookApiModel {
        return booksDataSource.getBooks(searchTerms = searchTerms)
    }

    suspend fun getThumbnails(bookId: String): GoogleBookRetrieveApiModel {
        return thumbnailDataSource.getThumbnails(bookId = bookId)
    }
}

File: SwiftUI-Bookshelf-App/Bookshelf app/Data/BooksRepository.swift
//
//  BooksRepository.swift
//  Bookshelf app
//
//  Created by Jaehwa Noh on 1/28/24.
//

import Factory

struct BooksRepository {
    @Injected(\.booksDataSource) var booksDataSource
    @Injected(\.bookThumbnailDataSource) var bookThumbnailDataSource
    
    func getBooks(searchTerms: String) async throws -> GoogleBookApiModel {
        try await booksDataSource.getBooks(searchTerms: searchTerms)
    }
    
    func getThumbnail(bookId: String) async throws -> GoogleBookRetrieveApiModel {
        try await bookThumbnailDataSource.getThumbnails(bookId: bookId)
    }
}
